# Исследование надежности заемщиков  
__Задача:__ На основе статистики о платёжеспособности клиентов исследовать влияет ли семейное положение и количество детей клиента на факт возврата кредита в срок
  
__План исследования:__  
- Знакомство с датасетом  
- Предобработка данных  
- Исследование данных  
- Общий вывод

## Знакомство с датасетом

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [6]:
data['days_employed'] = data['days_employed'].abs()

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [12]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [15]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [16]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [17]:
data['education'] = data['education'].str.lower()

In [18]:
data.duplicated().sum()

71

In [19]:
data = data.drop_duplicates()

### Категоризация данных

In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [25]:
children_pivot = data.pivot_table(index=['children'], values='debt', aggfunc=('count', 'sum'))
children_pivot['avg'] = (children_pivot['sum'] / children_pivot['count']).apply('{:.2%}'.format)
children_pivot = children_pivot[['count', 'avg']]
children_pivot

count    avg
children              
0         14091  7.54%
1          4808  9.23%
2          2052  9.45%
3           330  8.18%
4            41  9.76%
5             9  0.00%

**Вывод:** 
- Значимыми выборками являются выборки с 0, 1, 2 детьми, т.к. на них приходится ~98% всех заемщиков.
- Из приведенных данных можно сделать вывод, что платежная дисциплина заемщика ухудшается с ростом количества детей

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [45]:
def marriage_experience (status):
    if status != 'Не женат / не замужем' and status != 'гражданский брак' :
        return 'есть опыт'
    return 'нет опыта'
data['marriage_experience'] = data['family_status'].apply(marriage_experience)       
data.tail(30)
        
family_pivot = data.pivot_table(index=['marriage_experience'], values='debt', aggfunc=('count', 'sum'))
family_pivot['avg'] = (family_pivot['sum'] / family_pivot['count']).apply('{:.2%}'.format)
family_pivot = family_pivot[['count', 'avg']]
family_pivot

count    avg
marriage_experience              
есть опыт            14401  7.46%
нет опыта             6930  9.49%

**Вывод:** 
- Данные выборки возможно объединить по опыту официального брака. В результате получим две категории: "есть опыт" (входит: "женат/замужем", "вдовец/вдова", "в разводе") и "нет опыта" (входят все остальные). 
- Наблюдая данные две категории, можно предположить, что более дисциплинированные те заемщики, которые имеют опыт официального брака.


### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
income_pivot = data.pivot_table(index=['total_income_category'], values='debt', aggfunc=('count', 'sum'))
income_pivot['avg'] = (income_pivot['sum'] / income_pivot['count']).apply('{:.2%}'.format)
income_pivot = income_pivot[['count','avg']].sort_values(by='count', ascending=False)
income_pivot



count    avg
total_income_category              
C                      15921  8.50%
B                       5014  7.06%
D                        349  6.02%
A                         25  8.00%
E                         22  9.09%

**Вывод:** 
- Наиболее ценные выборки для анализа - категории С и В.
- Из данных следует, что  заемщики с доходами от 200 тыс. до 1 млн. руб. более дисциплинированы,   
нежели заемщики с доходами от 50 до 200 тыс. руб.

### Как разные цели кредита влияют на его возврат в срок?

In [25]:
targets_pivot = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=('count', 'sum'))
targets_pivot['avg'] = (targets_pivot['sum'] / targets_pivot['count']).apply('{:.2%}'.format)
targets_pivot = targets_pivot[['count', 'avg']].sort_values(by='count', ascending=False)
targets_pivot

count    avg
purpose_category                      
операции с недвижимостью  10751  7.26%
операции с автомобилем     4279  9.35%
получение образования      3988  9.25%
проведение свадьбы         2313  7.91%

**Вывод:** 
- Можно сделать вывод, что заемщики, берущие кредит на свадьбу и на операции с недвижимостью более дисциплинированы, чем заемщики, берущие кредит на получение образования и на покупку автомобиля.

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*  

- Человеческий фактор:  
  1. заемщик мог не указать данные о себе
  2. сотрудник банка не правильно заполнил анкеты
- Технический фактор:  
  1. сбои в работе оборудования или при передаче информации  
  2. различные форматы, которые используют системы отправляющие и получающие информацию

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:*   
В выборке могут находиться экстремальные значения, которые оказывают влияние на среднее значение. Медианное значение делит выборку ровно пополам (справа и слева находится равное количество элементов), игнорируя шум от экстремумов.

## Общий вывод.

  
Была получена выгрузка по заемщикам, имеющим историю получения кредита в данном банке. 
Выгрузка состоит из 21525 строк. Аномальные значения и дубликаты составили менее 1% от общего объема данных.  

Исходя из того, что в предоставленной таблице представлен формат "одним заемщик - одна цель кредита",  а также наличие менее 1% дубликатов, можно сделать допущение, что клеинты имели/имеют одним кредит в данном банке. 

Были проанализированы связи между возвратом кредита в срок и:  
- **количеством детей**   

В этой категории статистически значимыми являются выборки с 0, 1, 2 детьми. Из приведенных данных можно сделать вывод, что платежная дисциплина заемщика ухудшается с ростом количества детей. Разница между платежной дисциплиной заемщиков 0 и 2 детьми составляет почти 2%.  


- **семейным положением**  
  
Более значимые выборки в данной категории: "женат/замужем", "гражданский брак" и "не женат/не замужем". Объединив категории по наличию/отсутствию брачного опыта, можно предположить, более  дисциплинированные заемщики те, которые имеют опыт официального брака. Разница в дисциплине составила 2 %.

- **уровнем дохода**    
  
Наиболее ценные выборки для анализа - категории С и В, как наиболее многочисленные.
Из данных следует, что заемщики с более высокими доходами более ответствеено относятся к обязательствам. Разница в дисциплине между ближайшими категориями составляет 1.5%

- **целями кредита**  
  
В данной категории можно утверждать, что заемщики, берущие кредит на свадьбу и на операции с недвижимостью более дисциплинированы, чем заемщики, берущие кредит на получение образования и на покупку автомобиля. Разница в дисциплине составлет 1.5-2%%.

Все категории имеют закономерные перекосы, которые можно наблюдать в современном обществе:   
- количество детей редко превышает 1-2 ребенка  
- наиболее часто встречающаяся категория уровня дохода - средняя  
- вдовцов/вдов и разведенных значительно меньше остальных категорий  

Учитывая характер перекосов в категориях, для полноценной картины влияния различных факторов на платежную дисциплину заемщика, следует анализировать связи всех приведенных в таблице факторов с платежной дисциплиной, а так же принимать во внимание информацию из БКИ по наличию текущих кредитов, опыте кредитования, суммам и целям прошлых/текущих займов.